R/fct_SQLite.R

```diff
@@ -91,17 +91,99 @@ db_create <- function(
+    "db_version" = data.frame(version = db_version)
   )
   con <- get_db_connection(db_path)
-  for(i in names(new_data)){
+  db_add_tables(con, new_pk_data, idx_pk_cols, other_data)
+  cat("Finished writing to database\n\n")
+}
+
+#' Add new tables to DB
+#'
+#' @param con A DBI Connection to the SQLite DB
+#' @param pk_data A named list of data frames to add a primary key field to DB
+#'   table. Names will correspond to the DB table names.
+#' @param idx_cols A named list of the fields defining unique records for a
+#'   table. Names will correspond to the table to apply the index constraint.
+#' @param other_data A named list of other data frames to add to the DB. Names
+#'   will correspond to the DB table names.
+#'
+#' @keywords internal
+db_add_tables <- function(con, pk_data, idx_cols, other_data) {
+  for(i in names(pk_data)){
     cat("\nCreating new table: ", i,  "\n")
-    DBI::dbWriteTable(con, i, new_data[[i]])
+    db_add_primary_key(con, i, pk_data[[i]], idx_cols[[i]])
   }
-  cat("Finished writing to database\n\n")
+  for(i in names(other_data)){
+    cat("\nCreating new table: ", i,  "\n")
+    DBI::dbWriteTable(con, i, other_data[[i]])
+  }
+  cat("\nCreating log table: all_review_data_log\n")
+  db_add_log(con)
+}
+
+#' Add primary key field
+#' 
+#' @param con A DBI Connection to the SQLite DB
+#' @param name The table name
+#' @param value A data.frame to add to the table
+#' @param keys A character vector specifying which columns define a unique row
+#'   for the table. If `NULL`, no unique index will be created.
+#'   
+#' @keywords internal
+db_add_primary_key <- function(con, name, value, keys = NULL) {
+  fields <- c(id = "INTEGER PRIMARY KEY AUTOINCREMENT", DBI::dbDataType(con, value))
+  DBI::dbCreateTable(con, name, fields)
+  if (!is.null(keys)) {
+    rs <- DBI::dbSendStatement(con, sprintf("CREATE UNIQUE INDEX idx_%1$s ON %1$s (%2$s)", name, paste(keys, collapse = ", ")))
+    DBI::dbClearResult(rs)
+  }
+  DBI::dbAppendTable(con, name, value)
+}
+
+#' Add Logging Table
+#'
+#' Both creates the logging table and the trigger to update it for
+#' all_review_data.
+#' 
+#' @param con A DBI Connection to the SQLite DB
+#' 
+#' @keywords internal
+db_add_log <- function(con) {
+  DBI::dbCreateTable(con, "all_review_data_log",
+                     c(id = "INTEGER PRIMARY KEY AUTOINCREMENT", review_id = "INTEGER NOT NULL",
+                       edit_date_time = "CHAR", reviewed = "CHAR", comment = "CHAR", 
+                       reviewer = "CHAR", timestamp = "CHAR", status = "CHAR",
+                       dml_type = "CHAR NOT NULL", dml_timestamp = "DATETIME DEFAULT CURRENT_TIMESTAMP"))
+  rs <- DBI::dbSendStatement(con, paste(
+    "CREATE TRIGGER all_review_data_update_log_trigger",
+    "AFTER UPDATE ON all_review_data FOR EACH ROW",
+    "BEGIN",
+      "INSERT INTO all_review_data_log (",
+        "review_id, edit_date_time, reviewed, comment, reviewer, timestamp, status, dml_type",
```

Comment:
Wait I think it is that the review_id identifies one 'review' that can contain multiple unique rows, and therefore the numbers are different, right?

Reply:
The `review_id` matches up with `all_review_data$id` (if I may use this notation). The `id` is the primary key for the logging table so that records can more easily be queried from that table. The logging table does not contain the index/unique columns. If you wanted to look at the log, you would need to join to `all_review_data` to get the index/unique columns. Storing those columns in both spaces wouldn't be very efficient.

Reply:
> Wait I think it is that the review_id identifies one 'review' that can contain multiple unique rows, and therefore the numbers are different, right?

Yes, review_id identifies which review the logging record is referencing.